crates/sp-executor/src/lib.rs

```diff
@@ -202,6 +203,81 @@ pub struct BundleElectionParams {
     pub slot_probability: (u64, u64),
 }
 
+#[derive(Debug, Decode, Encode, TypeInfo, PartialEq, Eq, Clone)]
+pub enum VrfProofError {
+    /// Can not construct the vrf public_key/output/proof from the raw bytes.
+    VrfSignatureConstructionError,
+    /// Invalid vrf proof.
+    BadProof,
+}
+
+/// Verify the vrf proof generated in the bundle election.
+pub fn verify_vrf_proof(
+    public_key: &[u8],
+    vrf_output: &[u8],
+    vrf_proof: &[u8],
+    slot_randomness: &Blake2b256Hash,
+) -> Result<(), VrfProofError> {
+    let public_key = schnorrkel::PublicKey::from_bytes(public_key)
+        .map_err(|_| VrfProofError::VrfSignatureConstructionError)?;
+
+    public_key
+        .vrf_verify(
+            make_local_randomness_transcript(slot_randomness),
+            &VRFOutput::from_bytes(vrf_output)
+                .map_err(|_| VrfProofError::VrfSignatureConstructionError)?,
+            &VRFProof::from_bytes(vrf_proof)
+                .map_err(|_| VrfProofError::VrfSignatureConstructionError)?,
+        )
+        .map_err(|_| VrfProofError::BadProof)?;
+
+    Ok(())
+}
+
+#[derive(Debug, Decode, Encode, TypeInfo, PartialEq, Eq, Clone)]
+pub enum ReadBundleElectionParamsError {
+    /// Trie error.
+    TrieError,
+    /// The value does not exist in the trie.
+    MissingValue,
+    /// Failed to decode the value read from the trie.
+    DecodeError,
+}
+
+/// Returns the bundle election parameters read from the given storage proof.
+pub fn read_bundle_election_params(
+    storage_proof: StorageProof,
+    state_root: &H256,
+) -> Result<BundleElectionParams, ReadBundleElectionParamsError> {
+    use well_known_keys::{AUTHORITIES, SLOT_PROBABILITY, TOTAL_STAKE_WEIGHT};
+    use ReadBundleElectionParamsError::{DecodeError, MissingValue, TrieError};
```

Comment:
We generally don't use in-function imports unless they are platform-specific or otherwise conditionally compled. Also I don't think we used enum variant imports before either.